src/python/analyzer_executor/src/analyzer_executor_lib/analyzer_executor.py

```diff
@@ -54,157 +53,331 @@ def get(self, key):
-    MESSAGECACHE_PORT = int(os.environ["MESSAGECACHE_PORT"])
 
-    HITCACHE_ADDR = os.environ["HITCACHE_ADDR"]
-    HITCACHE_PORT = os.environ["HITCACHE_PORT"]
+class AnalyzerExecutor:
 
-    message_cache = redis.Redis(host=MESSAGECACHE_ADDR, port=MESSAGECACHE_PORT, db=0)
-    hit_cache = redis.Redis(host=HITCACHE_ADDR, port=int(HITCACHE_PORT), db=0)
+    # constants
+    CHUNK_SIZE_RETRY: int = 10
+    CHUNK_SIZE_DEFAULT: int = 100
 
+    # singleton
+    _singleton = None
 
-def parse_s3_event(s3, event) -> str:
-    bucket = event["s3"]["bucket"]["name"]
-    key = event["s3"]["object"]["key"]
-    return download_s3_file(s3, bucket, key)
+    def __init__(self, message_cache, hit_cache, chunk_size, is_local, logger):
+        self.message_cache = message_cache
+        self.hit_cache = hit_cache
+        self.chunk_size = chunk_size
+        self.is_local = is_local
+        self.logger = logger
 
+    @classmethod
+    def singleton(cls):
+        if not cls._singleton:
+            LOGGER.debug("initializing AnalyzerExecutor singleton")
+            is_local = bool(
+                os.getenv("IS_LOCAL", False)
+            )  # TODO move determination to grapl-common
 
-def download_s3_file(s3, bucket: str, key: str) -> str:
-    obj = s3.Object(bucket, key)
-    return obj.get()["Body"].read()
+            # If we're retrying, change the chunk size
+            is_retry = os.getenv("IS_RETRY", False)
+            if is_retry == "True":
+                chunk_size = cls.CHUNK_SIZE_RETRY
+            else:
+                chunk_size = cls.CHUNK_SIZE_DEFAULT
+
+            # Set up message cache
+            messagecache_addr = os.getenv("MESSAGECACHE_ADDR")
+            messagecache_port = os.getenv("MESSAGECACHE_PORT")
+            LOGGER.debug(f"{messagecache_addr}:{messagecache_port}")
+            try:
+                messagecache_port = int(messagecache_port)
+            except (TypeError, ValueError):
+                LOGGER.error(
+                    f"can't connect to redis, MESSAGECACHE_PORT couldn't cast to int"
+                )
+                messagecache_port = None
+
+            if messagecache_addr and messagecache_port:
+                LOGGER.debug(
+                    f"message cache connecting to redis at {messagecache_addr}:{messagecache_port}"
+                )
+                message_cache = redis.Redis(
+                    host=messagecache_addr, port=messagecache_port, db=0
+                )
+            else:
+                if bool(messagecache_addr) ^ bool(messagecache_port):
+                    LOGGER.error(
+                        f"message cache falling back to no-op cache (incomplete connection details)"
+                    )
+                else:
+                    LOGGER.debug(
+                        f"message cache falling back to no-op cache (missing connection details)"
+                    )
+                message_cache: EitherCache = NopCache()
```

Comment:
we should just fail if we can't connect to redis

Reply:
is there _any_ reason someone would want a null cache in prod?  i.e. should the connection info be completely mandatory, or should we allow a fallback NopCache?

Reply:
I think the connection should be mandatory - if we can't connect to Redis we have a pretty serious problem with out deployment. 

NopCache should strictly be a unit testing thing.